prompt_toolkit/shortcuts/progress_bar/base.py

```diff
@@ -359,6 +364,13 @@ def item_completed(self) -> None:
         self.items_completed += 1
         self.progress_bar.invalidate()
 
+    async def _remove_when_done_async(self):
+        def run_remove_when_done_thread() -> None:
+            return self.remove_when_done(self)
+
+        if await run_in_executor_with_context(run_remove_when_done_thread):
```

Comment:
I'm still thinking about whether this is the right thing to do.

By default, I'd like as much as possible for prompt_toolkit to have asynchronous APIs, and possibly create an synchronous abstraction on top of that. If we expect a blocking synchronous function here, then we exclude people that would like to include it in an async app.

Reply:
I borrowed almost all of this code from how validation and `ThreadedValidators` are implemented. I do see that the current implementation appears to have some issue with cleaning up tasks quickly upon completion (you'll see this if you run the a-lot-of-parallel-tasks.py example that I updated).

Would it be better to allow `remove_when_done` to be either synchronous or asynchronous and then it's up to the user to design any expensive async functions correctly (with `asyncio.sleep`, etc.)?